services/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantTypeBase.java

```diff
@@ -243,8 +254,15 @@ protected String getRequestedScopes() {
         }
 
         if (!validScopes) {
+            logger.debugf("Invalid scopes: %s. realm=%s client_id=%s", scope, realm.getName(), client.getClientId());
             event.error(Errors.INVALID_REQUEST);
-            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_SCOPE, "Invalid scopes: " + scope,
+            Set<String> allowedScopes = Stream.concat( //
```

Comment:
Here we collect all allowed scope names that can be requested (and represented in the token).